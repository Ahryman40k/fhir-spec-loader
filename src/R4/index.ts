export * from './SpecificationService';
export * from './ResourceTypeDescriptor';
export * from './PropertyDescriptor';
export * from './DescriptorIterator';
export * from './descriptor';
