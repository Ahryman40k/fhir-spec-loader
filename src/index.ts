export { IIterator, ISpecificationService } from './interfaces';

export { Loader, ArchiveKind, LoaderOptions  } from './loader';