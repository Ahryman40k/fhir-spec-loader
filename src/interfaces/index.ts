export {IIterator} from './IIterator';
export {ISpecificationService} from './ISpecificationService';