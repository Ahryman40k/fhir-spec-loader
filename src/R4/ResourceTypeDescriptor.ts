
import { ElementDefinition_TypeLike, ITypeDescriptor } from "../interfaces";

export class ResourceTypeDescriptor implements ITypeDescriptor {
    constructor(obj: ElementDefinition_TypeLike);

    constructor(private _raw: ElementDefinition_TypeLike) {
    }
    
    private _key?: string;
    get key(): string {
        if (!this._key) {
            /*if (this._raw.code && isArray(this._raw.code.extension)) {
                this._key = this._raw.code.extension[0].valueString;
            }
            else {
                this._key = this._raw.code as string;
            }*/
        }
        return this._key ? this._key : '';
    }

    /*
        get kind(): PropertyTypeKind {
            // array, enum, basic, resource
            return CodeSystem_PropertyTypeKind
        }*/
}
