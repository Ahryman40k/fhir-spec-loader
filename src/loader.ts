import { R4 } from '@ahryman40k/ts-fhir-types';
import { ISpecificationService } from './interfaces/ISpecificationService';
import * as fs from 'fs';
import * as http from 'http';
import * as tmp from 'tmp';
import * as unzipper from 'unzipper';

import { PathReporter } from 'io-ts/lib/PathReporter'
import { SpecificationService } from './SpecificationService';

export enum ArchiveKind {
    Zip = 'zip'
}

export type LoaderOptions = {
    format: ArchiveKind
}

function loadData(file: string): R4.IBundle {
    const raw = fs.readFileSync(file);

    let validation = R4.RTTI_Bundle.decode(JSON.parse(raw.toString()));
    if (validation.isLeft) {
        throw new Error(PathReporter.report(validation).join('/n'));
    }

    return <R4.IBundle>validation.value;
}

export class Loader {
    /* 
    * Extract files into temporary folder then call FromFiles()
    */
    static async FromArchive(filename: string, option: LoaderOptions = { format: ArchiveKind.Zip }): Promise<ISpecificationService> {

        const tmpDir = tmp.dirSync({ prefix: 'fhir-spec' });
        fs.createReadStream(filename)
            .pipe(unzipper.Extract({ path: tmpDir.name }));

        return this.FromFiles( fs.readdirSync(tmpDir.name ) );
    }

    static FromFiles(files: string[]): ISpecificationService {
        // TODO: Test files extension should be json
        try {
            return new SpecificationService(
                files.map(f => loadData(f))
            );

        } catch (ex) {
            throw ex;
        } finally {
            return new SpecificationService([]);
        }
    }

    static async FromWebsite(version: R4.StructureDefinitionFhirVersionKind): Promise<ISpecificationService> {
        if (version != R4.StructureDefinitionFhirVersionKind._400) {
            throw new Error('Required FHIR version is not supported by Loader class');
        }

        // Create temporary name
        const tmpFile = tmp.fileSync({ prefix: 'fhir-spec', postfix: '.zip' });
        const downloaded = fs.createWriteStream(tmpFile.name)

        // Download spec archive and save it with temparoray name
        await http.get("https://www.hl7.org/fhir/definitions.json.zip", response => {
            response.pipe(downloaded);
        });

        // Unzip archive
        return Loader.FromArchive(tmpFile.name);
    }
}