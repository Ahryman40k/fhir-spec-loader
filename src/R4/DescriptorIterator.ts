
import { IIterator, StructureDefinitionLike } from "../interfaces";
import { ItemDescriptor } from "./descriptor";


export class DescriptorIteratorDecorator implements IIterator<ItemDescriptor> {

    constructor(
        private _iter: IIterator<StructureDefinitionLike>
    ) {
    }

    next(): ItemDescriptor {
        if (!this._iter) {
            throw new Error('No more item to iterate');
        }


        return new ItemDescriptor(this._iter.next());
    }

    hasNext(): boolean {
        return this._iter ? this._iter.hasNext() : false;
    }
}



