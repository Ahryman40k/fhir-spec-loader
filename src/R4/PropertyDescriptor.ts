import { ElementDefinitionLike, IPropertyDescriptor, ITypeDescriptor } from "../interfaces";
import { FHIRCode2NativeType } from "./descriptor";

export class PropertyDescriptor implements IPropertyDescriptor {
    constructor(private _raw: ElementDefinitionLike) {
    }
    private _key?: string;
    get key(): string {
        if (!this._key) {
            const splitted = this._raw.id!.split('.');
            this._key = splitted.splice(1).join('.')
        }
        return this._key;
    }


    get description(): string {
        return this._raw.definition ?? '' + '\n' + this._raw.comment ?? ''
    }


    get optional(): boolean {
        return (this._raw.min ?? 0) === 0;
    }

    get isArray(): boolean {
        return this._raw.max ? this._raw.max === '*' : false;
    }

    get type(): ITypeDescriptor {

        const typeProperty = this._raw.type
        if ( typeProperty ) {
            const typeContainingCode = typeProperty.find( p => p.code !== undefined )
            // const typeContainingExts = typeProperty.find( p => p.extension !== undefined ) // This describe entity base type not property type

            if ( typeContainingCode ) {
                const code = typeContainingCode.code!
                return {
                    key: code,
                    baseType: code.startsWith( 'http://') ? FHIRCode2NativeType[code] : code
                }
            }
        }

        throw 'PropertyDescriptor::get type() is missing logic'
    }

    get inherited(): boolean {
        let result = false; 
        if ( this._raw.base) {
            result = this._raw.id !== this._raw.base.path;
        }
        
        return result;
    }

}
