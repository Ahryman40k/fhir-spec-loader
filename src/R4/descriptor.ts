
import { EntityType, IDescriptor, IPropertyDescriptor, ITypeDescriptor, StructureDefinitionLike } from "../interfaces";
import { PropertyDescriptor } from "./PropertyDescriptor";


export const FHIRCode2NativeType: Record<string, string> = {
    "http://hl7.org/fhirpath/System.String": 'string',
    "http://hl7.org/fhirpath/System.Date": 'string',
    "http://hl7.org/fhirpath/System.DateTime": 'string',
    "http://hl7.org/fhirpath/System.Time": 'string',
    "http://hl7.org/fhirpath/System.Decimal": 'number',
    "http://hl7.org/fhirpath/System.Integer": 'number',
    "http://hl7.org/fhirpath/System.Boolean": 'boolean',
}



function extractBaseType(definition: StructureDefinitionLike): ITypeDescriptor | undefined {
    switch (definition.kind) {
        
        case 'resource': {
            return { 
                key: definition.id,
                baseType: definition.baseDefinition!
            }
        }

        case 'primitive-type': {
            const value = definition.snapshot.element.find(element => element.id === `${definition.id}.value`)
            if (value) {
                const extensionGroups = value.type!.filter(t => t.extension !== undefined).map(t => t.extension!)
                const extensions = new Array<{ url: string, valueUri?: string, valueString?: string }>().concat(...extensionGroups)
                // const baseTypeExt = extensions.find(ext => ext.url === "http://hl7.org/fhir/StructureDefinition/structuredefinition-fhir-type")
                const code =  value.type!.find( t => t.code !== undefined )?.code ?? ''
                const regExExt = extensions.find(ext => ext.url === "http://hl7.org/fhir/StructureDefinition/regex")
                return {
                    regex: regExExt ? regExExt.valueString : undefined,
                    key: value.id,
                    baseType: FHIRCode2NativeType[code]
                }
            }
        }
        case 'complex-type':
        default:
            return undefined
    }
}


function extractProperties(definition: StructureDefinitionLike): Map<string, IPropertyDescriptor> {
    switch (definition.kind) {
        case 'complex-type':
        case 'resource': {
            const ps = definition.snapshot.element.slice(1).map(e => new PropertyDescriptor(e));
            return new Map(ps.map(p => [p.key, p]));
        }

        case 'primitive-type':
        default:
            return new Map<string, IPropertyDescriptor>()
    }
}


export class ItemDescriptor implements IDescriptor {
    constructor(private _raw: StructureDefinitionLike) {
    }

    get key(): string {
        return this.
        _raw.id;
    }

    get kind(): EntityType {
        return this._raw.kind;
    }

    get version(): string {
        return this._raw.fhirVersion;
    }

    private _properties?: Map<string, IPropertyDescriptor>;
    get properties(): Map<string, IPropertyDescriptor> {
        if (!this._properties) {
            this._properties = extractProperties(this._raw);
        }
        return this._properties!;
    }

    private _constraints?: string[];
    get constraints(): string[] {
        if (!this.constraints) {
            const baseElement = this._raw.snapshot.element.filter(element => !element.id.includes('.'))
            if (baseElement) {
                const keys = baseElement.map(base => {
                    if (base.constraint) {
                        return base.constraint.map(constraint => constraint.key)
                    }

                    return undefined
                })

                const filteredKeys = keys.filter((item): item is string[] => Boolean(item))
                this._constraints = new Array<string>().concat(...filteredKeys)
            }
        }

        return this._constraints!
    }

    get searchParameters(): Map<string, IPropertyDescriptor> {
        throw new Error('Not yet implemented');
    }

    private _baseType?: ITypeDescriptor;
    get inheritFrom(): ITypeDescriptor {
        if (!this._baseType) {
            this._baseType = extractBaseType(this._raw)
        }

        return this._baseType!
    }

    get description(): string {
        return this._raw.description ? this._raw.description : '';
    }
}
