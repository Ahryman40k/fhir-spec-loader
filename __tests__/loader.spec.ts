

import { StructureDefinitionFhirVersionKind } from '@ahryman40k/ts-fhir-types/lib/R4';
import * as path from 'path';
import { FromArchive, FromFiles, FromWebsite } from '../src/loader';

describe('Loaders', () => {
    describe('Loading from files', () => {

        test("should provide a specification service", () => {
            const service = FromFiles([
                path.join(__dirname, '../../../R4/definition/profiles-resources.json'),
            ]);
            expect(service).toBeDefined();
            // @ts-ignore
            expect(service['_rawSpec'].length).toBe(1);
        });

        test("should provide empty service if no files", () => {
            const service = FromFiles([]);
            expect(service).toBeDefined();
            // @ts-ignore
            expect(service['_rawSpec'].length).toBe(0);
        });



    })

    describe('Loading from archive', () => {

        test("should provide a specification service", () => {

            //  expect(T.decode(value)._tag).toBe('Right');
        });

        test("should fail if archive isn't zip file", () => {

            return FromArchive(path.join(__dirname, 'loader.spec.ts'))
                .then(service => {
                    expect(true).toBe(false);
                }).catch(err => {
                    expect(err).toBeDefined();
                });
        });

    })

    describe('Loading from website', () => {

        test("should provide a specification service", () => {

            return FromWebsite(StructureDefinitionFhirVersionKind._400).then(service => {
                expect(service).toBeDefined();
                // @ts-ignore
                expect(service['_rawSpec'].length).toBe(0);
            })
        }, 100000);


        test("should fail if version is different from R4", () => {

            return FromWebsite(StructureDefinitionFhirVersionKind._350).then(service => {
                expect(true).toBe(false);
            }).catch(err => {
                expect(err).toMatchSnapshot();
                expect(err).toBeDefined();
            })
        });
    })
})