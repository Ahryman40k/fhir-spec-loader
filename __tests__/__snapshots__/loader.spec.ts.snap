// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loaders Loading from website should fail if version is different from R4 1`] = `[Error: Required FHIR version is not supported by Loader class]`;
